### PR TITLE
Add stronger TLS check for Gossip in Cluster mode

### PR DESCRIPTION
**Pull request by msft-paddy14**
This PR resolves issue #391.

It's a small change that uses ClusterTlsClientTargetHost  and validates the incoming remote cert (during Gossip CLUSTER MEET call) and ensures that we're attempting to join a node that was created as part of current cluster setup. This can prevent issues of unknown clusters getting connected when multiple caches might be running in same IP pool. Previously, no such validation was there and we did an unconditional handshake approval (when ClientCertRequired is false). This will enforce the checks for only gossip where the server node acts as a client

### CODE REVIEW

**Comment by msft-paddy14** [comment]
@lmaas / @vazois  gentle ping for reviewing this change

File: libs/server/TLS/GarnetTlsOptions.cs [hunk 1]
```diff
@@ -161,21 +161,69 @@ SslClientAuthenticationOptions GetSslClientAuthenticationOptions()
             };
         }
 
+
+        /// <summary>
+        /// Callback to verify the TLS certificate
+        /// </summary>
+        /// <param name="issuerCertificatePath"></param>
```
Comment (lmaas) [inline libs/server/TLS/GarnetTlsOptions.cs:168]: NIT: Incomplete documentation comments. (I know the current code is not very consistently documented either, but we should try to have complete commends for all new functions)

File: libs/server/TLS/GarnetTlsOptions.cs [hunk 2]
```diff
@@ -161,21 +161,69 @@ SslClientAuthenticationOptions GetSslClientAuthenticationOptions()
...
+            return (object _, X509Certificate certificate, X509Chain __, SslPolicyErrors sslPolicyErrors)
+                => (sslPolicyErrors == SslPolicyErrors.None) || (sslPolicyErrors == SslPolicyErrors.RemoteCertificateChainErrors
+                    && certificate is X509Certificate2 certificate2
+                    && ValidateCertificateIssuer(certificate2, issuer));
+        }
+
+        /// <summary>
+        /// Returns issuer certificate in X509Certificate2 format.
+        /// </summary>
+        /// <param name="issuerCertificatePath">path to issuer cert.</param>
```
Comment (lmaas) [inline libs/server/TLS/GarnetTlsOptions.cs:224]: NIT: I'd clean up the comment a bit.

```suggestion
        /// <summary>
        /// Loads an issuer X.509 certificate using its file name.
        /// </summary>
        /// <param name="issuerCertificatePath">The path of the certificate file.</param>
```

File: libs/server/TLS/GarnetTlsOptions.cs [hunk 3]
```diff
@@ -161,21 +161,69 @@ SslClientAuthenticationOptions GetSslClientAuthenticationOptions()
...
+                   && ValidateCertificateName(certificate2, targetHostName)
+                   && ValidateCertificateIssuer(certificate2, issuer));
+        }
+
+        /// <summary>
+        /// Validates certificate subject name by looking into DNS name property (preferred), if missing it falls back to
+        /// legacy SimpleName. The input certificate subject should match the expected host name provided in server config.
+        /// </summary>
+        /// <param name="certificate2"> remote certificate to validate.</param>
+        /// <param name="targetHostName"> target host name configured. </param>
```
Comment (lmaas) [inline libs/server/TLS/GarnetTlsOptions.cs:187]:
```suggestion
        /// <param name="certificate2">The remote certificate to validate.</param>
        /// <param name="targetHostName">The expected target host name. </param>
```

File: libs/server/TLS/GarnetTlsOptions.cs [hunk 4]
```diff
@@ -161,21 +161,69 @@ SslClientAuthenticationOptions GetSslClientAuthenticationOptions()
...
+                => (sslPolicyErrors == SslPolicyErrors.None) || (sslPolicyErrors == SslPolicyErrors.RemoteCertificateChainErrors
+                    && certificate is X509Certificate2 certificate2
+                    && ValidateCertificateIssuer(certificate2, issuer));
+        }
+
+        /// <summary>
+        /// Returns issuer certificate in X509Certificate2 format.
+        /// </summary>
+        /// <param name="issuerCertificatePath">path to issuer cert.</param>
+        X509Certificate2 GetCertIssuer(string issuerCertificatePath)
```
Comment (lmaas) [inline libs/server/TLS/GarnetTlsOptions.cs:225]: NIT: I'd rename the function to `GetCertificateIssuer` for consistency.

**Review by lmaas — APPROVED** [review]
The code looks solid! I have added a few suggestions for minor code cleanup.

File: libs/server/TLS/GarnetTlsOptions.cs [hunk 5]
```diff
@@ -161,21 +161,69 @@ SslClientAuthenticationOptions GetSslClientAuthenticationOptions()
...
+                   && ValidateCertificateName(certificate2, targetHostName)
+                   && ValidateCertificateIssuer(certificate2, issuer));
+        }
+
+        /// <summary>
+        /// Validates certificate subject name by looking into DNS name property (preferred), if missing it falls back to
+        /// legacy SimpleName. The input certificate subject should match the expected host name provided in server config.
+        /// </summary>
+        /// <param name="certificate2">the remote certificate to validate.</param>
+        /// <param name="targetHostName">The expected target host name. </param>
```
Comment (lmaas) [inline libs/server/TLS/GarnetTlsOptions.cs:187]: NIT: Typo

```suggestion
        /// <param name="certificate2">The remote certificate to validate.</param>
        /// <param name="targetHostName">The expected target host name.</param>
```

File: libs/server/TLS/GarnetTlsOptions.cs [hunk 6]
```diff
@@ -161,21 +161,69 @@ SslClientAuthenticationOptions GetSslClientAuthenticationOptions()
...
+            return subjectName.Equals(targetHostName, StringComparison.InvariantCultureIgnoreCase);
+        }
+
+        /// <summary>
+        /// Callback to verify the TLS certificate
+        /// </summary>
+        /// <param name="issuerCertificatePath">The path to issuer certificate file.</param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentException"></exception>
+        /// <exception cref="ArgumentNullException"></exception>
```
Comment (lmaas) [inline libs/server/TLS/GarnetTlsOptions.cs:205]: NIT: I suggest completing these documentation comments, while we are at it.

File: libs/server/TLS/GarnetTlsOptions.cs [hunk 7]
```diff
@@ -161,21 +161,69 @@ SslClientAuthenticationOptions GetSslClientAuthenticationOptions()
...
+            return (object _, X509Certificate certificate, X509Chain __, SslPolicyErrors sslPolicyErrors)
+                => (sslPolicyErrors == SslPolicyErrors.None) || (sslPolicyErrors == SslPolicyErrors.RemoteCertificateChainErrors
+                    && certificate is X509Certificate2 certificate2
+                    && ValidateCertificateIssuer(certificate2, issuer));
+        }
+
+        /// <summary>
+        /// Loads an issuer X.509 certificate using its file name.
+        /// </summary>
+        /// <param name="issuerCertificatePath"> The path to issuer certificate file.</param>
```
Comment (lmaas) [inline libs/server/TLS/GarnetTlsOptions.cs:224]: NIT: Unnecessary whitespace.

```suggestion
        /// <param name="issuerCertificatePath">The path to issuer certificate file.</param>
```

**Review by lmaas — APPROVED** [review]
Great work! I think we can go ahead and merge this PR.
